### ISSUE TITLE
Modify `LogicalPlan::apply_with_subqueries()` and `LogicalPlan::visit_with_subqueries()` to return subqueries

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#10678]
### Is your feature request related to a problem or challenge?

in https://github.com/apache/datafusion/pull/10543 @peter-toth allowed the `TreeNode::apply` and `TreeNode::visit` APIs to return references 

However `LogicalPlan::apply_with_subqueries()` and `LogicalPlan::visit_with_subqueries()`, that are similar to TreeNode's base APIs but provide subquery support, weren't changed as well stricter easily. 

This is because in LogicalPlan::apply_expressions() and LogicalPlan::apply_subqueries() we create temporary Expr::eq, Expr::Column and LogicalPlan::Subquery objects that are not compatible with the root treenode's lifetime.

### Describe the solution you'd like

Change https://github.com/apache/datafusion/blob/c095fee4a1675f37385656c110ae7aa553fb1a55/datafusion/expr/src/logical_plan/tree_node.rs#L777-L780

to be 

```rust
    fn apply_with_subqueries<'n, F: FnMut(&'n Self) -> Result<TreeNodeRecursion>>(
        &'n self,
        mut f: F,
    ) -> Result<TreeNodeRecursion> {
```

### Describe alternatives you've considered

@peter-toth  says

> If we wanted to make the LogicalPlan visitor APIs stricter we would need further changes in LogicalPlan::Join, LogicalPlan::Unnest, LogicalPlan::Extension, Expr::Exists, Expr::InSubquery, Expr::ScalarSubquery enums, but I don't think we need those stricter APIs currently...

**Comment by alamb** [comment]
I don't think this is useful anymore, so let's close it. We can revisit how apply_with_subqueries works if needed for some specific usecase

**Comment by peter-toth** [comment]
Yeah, I remember that inconsistency in the API. It's tricky to fix. But I think you are right and we can revisit it once we actually need it.

**Comment by alamb** [comment]
👋  @peter-toth  -- hope you are doing well

**Comment by peter-toth** [comment]
I'm doing good, hope you too! 👋 I'm sure we will work together more often in the future! 😉
